### PR TITLE
Feature manage library bagfix

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: sql_ddl/db.sql [hunk 1]
```diff
@@ -18,3 +19,35 @@ create table reset_token
...
     user_id integer
 );
+
+create table actions
+(
+    id             serial primary key not null,
+    title          varchar(10)        not null,
+    description    varchar(50),
+    idCompound     integer,
+    idTestScenario integer,
```
Comment (henchamax) [inline sql_ddl/db.sql]: 1. Please, use the same naming convention as previously. 
2. Foreign keys ?

File: sql_ddl/db.sql [hunk 2]
```diff
@@ -18,3 +19,35 @@ create table reset_token
...
+    status         action_status not null
+
+);
+create table compounds
+(
+    id serial primary key  not null ,
+    title varchar(10),
+    description    varchar(50),
+    idTestScenario integer,
```
Comment (henchamax) [inline sql_ddl/db.sql]: 1. Please, use the same naming convention as previously.
2. Foreign keys ?

File: src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+public class ActionsController {
+    private final ActionService actionService;
+
+    @Autowired
+    public ActionsController(ActionService actionService) {
+        this.actionService = actionService;
+    }
+
+    @GetMapping("/actions")
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java]: You can move this prefix to controller level

File: sql_ddl/db.sql [hunk 4]
```diff
@@ -18,3 +19,35 @@ create table reset_token
...
+    status         action_status not null
+);
+create table lib_act_compound
+(
+    id_library integer,
+    id_compound integer,
+    id_action integer
```
Comment (henchamax) [inline sql_ddl/db.sql:45]: 1. Foreign keys ?
2. Primary key is missing

File: sql_ddl/db.sql [hunk 5]
```diff
@@ -18,3 +19,35 @@ create table reset_token
...
+);
+create table library
+(
+    id serial primary key  not null,
+    description varchar(50),
+    name varchar(10) not null,
+    createdate timestamp not null
```
Comment (henchamax) [inline sql_ddl/db.sql]: Naming

File: src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    @PutMapping("actions/change-status")
+    public Action changeStatus(@RequestBody ChangeActionStatusDTO changeActionStatusDTO) {
+        return actionService.changeStatus(changeActionStatusDTO.getStatusToChange(),
+                changeActionStatusDTO.getId());
+    }
+
+    @DeleteMapping("/actions")
+    public ResponseEntity<ActionStatusResponse> deleteAction(
+            @RequestHeader("Authorisation") String token,
+            @RequestBody DeleteActionDTO deleteActionDTO) {
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java]: Request body in DELETE http method ?

File: src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @PostMapping("/actions")
+    public ResponseEntity<CreateActionResponse> createAction(@RequestParam(name = "id") int idLibrary, @RequestBody Action actionDTO) {
+        actionService.createAction(idLibrary, actionDTO);
+        return ResponseEntity.ok(new CreateActionResponse("Success"));
+
+    }
+
+    @GetMapping("actions/by-id/{id}")
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java]: Please, change url

File: src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    public Action getActionsByTitle(@PathVariable int id) {
+        return actionService.getActionById(id);
+    }
+
+
+    @PutMapping("actions")
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java]: Does this endpoint work ?

File: src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        this.actionService = actionService;
+    }
+
+    @GetMapping("/actions")
+    public List<Action> getAllActions() {
+        return actionService.selectAllActions();
+    }
+
+    @PostMapping("/actions")
+    public ResponseEntity<CreateActionResponse> createAction(@RequestParam(name = "id") int idLibrary, @RequestBody Action actionDTO) {
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java]: Can we create action without library ?

File: src/main/java/com/ncedu/cheetahtest/controller/compound/CompoundController.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api")
+@CrossOrigin(origins = "${frontend.ulr}")
+
+public class CompoundController {
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/compound/CompoundController.java:15]: The same questions as for ActionsController

File: src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        actionService.createAction(idLibrary, actionDTO);
+        return ResponseEntity.ok(new CreateActionResponse("Success"));
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/action/ActionsController.java]: Better return created action

File: src/main/java/com/ncedu/cheetahtest/controller/library/LibraryController.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api")
+@CrossOrigin(origins = "${frontend.ulr}")
+
+public class LibraryController {
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/library/LibraryController.java:23]: The same questions as for ActionsController

File: src/main/java/com/ncedu/cheetahtest/dao/compound/CompoundDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        String sql = CREATE_COMPOUND;
+        jdbcTemplate.update(
+                sql,
+                compound.getTitle(),
+                compound.getDescription(),
+                compound.getIdTestScenario(),
+                compound.getStatus()
+        );
+        sql = SELECT_CURRVAL_COMPOUNDS_ID;
+        List<Integer> currIndex = jdbcTemplate.query(sql, new CurrentValueRowMapper());
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/dao/compound/CompoundDaoImpl.java:67]: What if someone create another compound before this query ?

File: src/main/java/com/ncedu/cheetahtest/dao/compound/CompoundDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+                preparedStatement -> {
+                    preparedStatement.setString(1, title);
+                    preparedStatement.setInt(2, idLibrary);
+                },
+                new CompoundRowMapper()
+        );
+    }
+
+    @Override
+    public int createCompound(Compound compound) {
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/dao/compound/CompoundDaoImpl.java:57]: Better return created compound

File: src/main/java/com/ncedu/cheetahtest/service/action/ActionServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+        this.actionDao = actionDao;
+        this.libActCompoundDao = libActCompoundDao;
+    }
+
+    @Override
+    public void createAction(int idLibrary, Action actionDTO) {
+
+        if (actionDTO.getTitle() == null || (!actionDTO.getStatus().equals("active") &&
+                !actionDTO.getStatus().equals("inactive"))) {
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/service/action/ActionServiceImpl.java]: Please, use Yoda conditions

File: src/main/java/com/ncedu/cheetahtest/service/action/ActionService.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+public interface ActionService {
+
+    void createAction(int idLibrary, Action actionDTO);
+    List<Action> selectAllActions();
+    Action getActionById(int id);
+    List<Action> getActiveActionsByTitle(int idLibrary,String title);
+    Action editAction(Action actionDTO);
+    Action changeStatus(String status, int id);
+    boolean isAdmin(String jwtToken);
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/service/action/ActionService.java]: Why isAdmin service in ActionServise ?

File: src/main/java/com/ncedu/cheetahtest/service/compound/CompoundServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+import com.ncedu.cheetahtest.exception.managelibraries.RightsPermissionException;
+import com.ncedu.cheetahtest.exception.managelibraries.UnproperInputException;
+import org.apache.tomcat.util.codec.binary.Base64;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+public class CompoundServiceImpl implements CompoundService {
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/service/compound/CompoundServiceImpl.java:16]: Same as for ActionService

File: src/main/java/com/ncedu/cheetahtest/service/library/LibraryService.java [hunk 18]
```diff
@@ -0,0 +1,14 @@
+package com.ncedu.cheetahtest.service.library;
+
+import com.ncedu.cheetahtest.entity.library.Library;
+
+import java.util.List;
+
+public interface LibraryService {
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/service/library/LibraryService.java:7]: Same as for ActionService

File: src/main/java/com/ncedu/cheetahtest/controller/library/LibraryController.java [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        this.compoundService = compoundService;
+    }
+
+
+    @GetMapping("/libraries")
+    public LibraryPaginationResponseBody getLibraryByName(
+            @RequestParam(name = "title") String title,
+            @RequestParam(name = "size") int size,
+            @RequestParam(name = "page") int page) {
+        List<Library> libraries = libraryService.getLibrariesByName(title);
```
Comment (henchamax) [inline src/main/java/com/ncedu/cheetahtest/controller/library/LibraryController.java:41]: Pagination should work in another way.

File: sql_ddl/db.sql (hunk 1 above)
Comment (HellaxA) [inline sql_ddl/db.sql]: You should name these fields like this: compound_id, test_scenario_id

**Comment by GorovDimon11** [comment]
Viewed

**Review by KotyaraSingleCat — APPROVED** [review]

**Comment by HellaxA** [comment]
ok

**Review by RomaSliva — APPROVED** [review]

**Review by danilos1 — APPROVED** [review]

**Pull request merged**